core/src/sumeragi/fault.rs

```diff
@@ -39,865 +37,1333 @@ impl FaultInjection for NoFault {
-        if header.is_genesis() && self.block_height == 0 {
-            if let Some(genesis_topology) = &header.genesis_topology {
-                info!("Using network topology from genesis block");
-                return genesis_topology.clone();
+                let mut vote_count = 0;
+                let mut peer_has_voted = vec![false; validating_peers.len()];
+                let mut peer_signatures = Vec::new();
+                for signature in signatures_on_this_block {
+                    for i in 0..validating_peers.len() {
+                        if *signature.public_key() == validating_peers[i].public_key {
+                            if !peer_has_voted[i] {
+                                peer_has_voted[i] = true;
+                                vote_count += 1;
+                                peer_signatures.push(signature.clone());
+                            }
+                            break;
+                        }
+                    }
+                }
+
+                vote_count += 1; // We are also voting for this block.
+                if vote_count >= state_machine_guard.current_topology.min_votes_for_commit() {
+                    let mut block = voting_block_option.unwrap().block;
+                    voting_block_option = None;
+
+                    block.as_mut_v1().signatures = peer_signatures
+                        .into_iter()
+                        .map(SignatureOf::transmute)
+                        .collect();
+                    let block = block
+                        .sign(sumeragi.key_pair.clone())
+                        .expect("Why should signing fail?");
+
+                    assert!(
+                        block.as_v1().signatures.len()
+                            >= state_machine_guard.current_topology.min_votes_for_commit()
+                    );
+
+                    info!(
+                        %voting_block_hash,
+                        "Block reached required number of votes",
+                    );
+
+                    sumeragi.broadcast_msg(
+                        BlockCommitted::from(block.clone()),
+                        &state_machine_guard.current_topology,
+                    );
+                    block_commit(sumeragi, block, &mut state_machine_guard);
+                }
+            }
+
+            if voting_block_option.is_some()
+                && Instant::now() > instant_at_which_we_should_have_committed
+            {
+                trace!("Suspecting validating peers for not voting for block.");
+                request_view_change(
+                    sumeragi,
+                    &mut state_machine_guard,
+                    &mut view_change_proof_chain,
+                    current_view_change_index,
+                );
+                instant_at_which_we_should_have_committed += sumeragi.commit_time;
             }
         }
+    }
+}
```

Comment:
I'm unable to read this function. This is the **main** part of the whole *Iroha*. This part should be audited by other companies and programmers.
This is not *ready for lts* state. We need a great refactor here

Reply:
It is the main part. And it's only 1000 lines, that's remarkably small in my opinion. Adding more indirection and increasing sloc count doesn't make it easier to understand in my opinion. I wrote this code following principles laid out here: http://number-none.com/blow/blog/programming/2014/09/26/carmack-on-inlined-code.html